#Домашнє завдання
В якості домашнього завдання вам пропонується створити рекурентну нейронну мережу за допомогою механізмів Keras, яка буде класифікувати рецензії із датасету imdb.

На відміну від прикладу в модулі 9 ми використаємо рекурентну нейронну мережу. Поекспериментуйте з будовою мережі - RNN, LSTM, двостороння та глибока.

Порівняйте результати та зробіть висновки.

##LSTM (Long Short-Term Memory)

In [12]:
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing import sequence
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense

In [13]:
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=5000)

In [14]:
max_words = 500
x_train = sequence.pad_sequences(x_train, maxlen=max_words)
x_test = sequence.pad_sequences(x_test, maxlen=max_words)

In [15]:
model = Sequential()
model.add(Embedding(5000, 32, input_length=max_words))
model.add(LSTM(100))
model.add(Dense(1, activation='sigmoid'))

In [16]:
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [17]:
model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=5, batch_size=64)

Epoch 1/5
391/391 [==============================] - 456s 1s/step - loss: 0.5155 - accuracy: 0.7409 - val_loss: 0.3812 - val_accuracy: 0.8387
Epoch 2/5
391/391 [==============================] - 451s 1s/step - loss: 0.3231 - accuracy: 0.8662 - val_loss: 0.3353 - val_accuracy: 0.8594
Epoch 3/5
391/391 [==============================] - 447s 1s/step - loss: 0.2505 - accuracy: 0.9012 - val_loss: 0.3234 - val_accuracy: 0.8647
Epoch 4/5
391/391 [==============================] - 449s 1s/step - loss: 0.2214 - accuracy: 0.9142 - val_loss: 0.3365 - val_accuracy: 0.8652
Epoch 5/5
391/391 [==============================] - 439s 1s/step - loss: 0.2003 - accuracy: 0.9236 - val_loss: 0.3567 - val_accuracy: 0.8533


In [9]:
_, accuracy = model.evaluate(x_test, y_test)

782/782 [==============================] - 99s 126ms/step - loss: 0.3339 - accuracy: 0.8728


In [10]:
print('Accuracy:', accuracy)

Accuracy: 0.872759997844696


##Recurrent neural networks (RNN)

In [18]:
from tensorflow.keras.layers import SimpleRNN

In [19]:
model_rnn = Sequential()
model_rnn.add(Embedding(5000, 32, input_length=max_words))
model_rnn.add(SimpleRNN(100))
model_rnn.add(Dense(1, activation='sigmoid'))

In [20]:
model_rnn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])


In [21]:
model_rnn.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=5, batch_size=64)

Epoch 1/5
391/391 [==============================] - 130s 328ms/step - loss: 0.5981 - accuracy: 0.6632 - val_loss: 0.4637 - val_accuracy: 0.8034
Epoch 2/5
391/391 [==============================] - 128s 328ms/step - loss: 0.4076 - accuracy: 0.8204 - val_loss: 0.5119 - val_accuracy: 0.7496
Epoch 3/5
391/391 [==============================] - 128s 327ms/step - loss: 0.5215 - accuracy: 0.7241 - val_loss: 0.6246 - val_accuracy: 0.6266
Epoch 4/5
391/391 [==============================] - 131s 335ms/step - loss: 0.5823 - accuracy: 0.6843 - val_loss: 0.5836 - val_accuracy: 0.6837
Epoch 5/5
391/391 [==============================] - 129s 331ms/step - loss: 0.4136 - accuracy: 0.8156 - val_loss: 0.4592 - val_accuracy: 0.8008


In [22]:
_rnn, accuracy_rnn = model_rnn.evaluate(x_test, y_test)

782/782 [==============================] - 34s 43ms/step - loss: 0.4592 - accuracy: 0.8008


In [23]:
print('Accuracy RNN:', accuracy_rnn)

Accuracy RNN: 0.8007599711418152


###Висновок
Порівнявши результати двох рекурентних шарів, LSTM та SimpleRNN, на датасеті IMDb, отримали такі результати:

- LSTM: Accuracy = 0.872759997844696
- SimpleRNN: Accuracy = 0.8007599711418152
За результатами можна зробити наступні висновки:

1. LSTM показує кращі результати точності класифікації порівняно з SimpleRNN на датасеті IMDb. Це може бути пов'язано з тим, що LSTM здатна краще управляти довгостроковими залежностями, що часто зустрічаються в текстових даних.

2. Використання LSTM у рекурентних нейронних мережах може допомогти покращити точність класифікації текстових даних, таких як рецензії. Він зберігає інформацію про довгострокові залежності у тексті, що дозволяє краще розуміти контекст та зміст рецензій.

3. SimpleRNN, хоча менш потужний за LSTM, все ще може бути корисним у випадках, де важлива швидкість навчання та ефективність моделі. Він може дати задовільні результати для деяких простіших текстових завдань.

Загалом, використання LSTM у рекурентних нейронних мережах показує переваги у точності класифікації текстових даних, особливо у випадку аналізу рецензій.